### PR TITLE
Breaking: Implement Redux

### PR DESCRIPTION
**Pull request by omgaz**
- Move form and select examples to example components
- add redux, react-redux dependencies (non exportable)
- Add redux examples, tests and linting
- Upgrade package dependencies for react-select

### CODE REVIEW

File: src/examples/selects.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
+
```
Comment (omgaz) [inline src/examples/selects.js]: extra line

File: src/helpers/reactSelectAllowCreateOverride.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
+/**
+ * As react-select 1 is still in beta the allowCreate flag isn't supported.
```
Comment (omgaz) [inline src/helpers/reactSelectAllowCreateOverride.js]: v1

File: src/components/Main.js [hunk 3]
```diff
@@ -1,44 +1,55 @@
...
+        shown: true,
+        results: action.results
+      };
+    default:
+      return state;
+  }
+};
+
+const reducer = combineReducers({
+  exampleFormReducer: exampleFormReducer,
```
Comment (camjc) [inline src/components/Main.js]: `: exampleFormReducer` :fire:

File: src/components/adslotUi/ReduxSelect.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import React from 'react';
+import Select from 'react-select';
+import { filterOptions } from '../../helpers/reactSelectAllowCreateOverride';
+
+const ReduxSelect = (props) => {
+  const {defaultValue, onBlur, value, allowCreate} = props;
+  const val = value || defaultValue;
+  const onBlurBound = () => onBlur(value);
+
+  let allowCreateOverride = allowCreate? { filterOptions } : {};
```
Comment (camjc) [inline src/components/adslotUi/ReduxSelect.js]: const. Linting should pick this up?

File: src/components/Main.js [hunk 5]
```diff
@@ -22,30 +21,46 @@ import {
...
-  { label: 'Strawberry', value: 'strawberry' },
-  { label: 'Caramel', value: 'caramel' },
-  { label: 'Cookies and Cream', value: 'cookiescream' },
-  { label: 'Peppermint', value: 'peppermint' },
-];
+const formReducer = (state = {}, action) => {
+  switch (action.type) {
+    case 'save':
+      alert(JSON.stringify(action.formValues));
+      return _.assign({}, state, action.formValues);
```
Comment (omgaz) [inline src/components/Main.js]: Use immutable JS

File: src/components/Main.js [hunk 6]
```diff
@@ -22,30 +21,46 @@ import {
...
+    case 'update':
+      alert(JSON.stringify(action.values));
+      return _.assign({}, state, action.values);
+    default:
+      return state;
+  }
+};
+
+const visibilityReducer = (state = { isSubmitting: false }, action) => {
```
Comment (omgaz) [inline src/components/Main.js]: and this immute

File: src/components/Main.js [hunk 7]
```diff
@@ -22,30 +21,46 @@ import {
...
+      return state;
+  }
+};
+
+const reducer = combineReducers({
+  form: formReducer,
+  visibility: visibilityReducer,
+});
+
+const store = createStore(reducer); // One store per app.
```
Comment (omgaz) [inline src/components/Main.js]: comment is kinda useless

File: src/examples/forms.js [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+          helpText="Helpful text."
+          label="Form Group"
+          placeholder="5.00"
+        />
+
+        <div className="form-group">
+          <label htmlFor="exampleTextInput" className="control-label col-xs-3">Text input</label>
+          <div className="col-xs-9">
+            <div className="input-group col-xs-6">
+              <input type="text" className="form-control" id="exampleTextInput" placeholder="Text input" />
```
Comment (omgaz) [inline src/examples/forms.js]: onChange

File: src/examples/forms.js [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+  updateValues: (values) => {
+    dispatch({ type: 'update', values });
+  },
+
+  validateAndSave: (values) => {
+    dispatch({ type: 'submitting', isSubmitting: true });
+    return new Promise((resolve) => {
+      setTimeout(() => {
+        dispatch({ type: 'submitting', isSubmitting: false });
+        if (isNaN(Number(values.exampleTextWithAddon))) {
```
Comment (omgaz) [inline src/examples/forms.js]: doesnt exist

File: src/examples/forms.js [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+    });
+  },
+});
+
+const mapStateToProps = (state) => ({
+  submitting: state.visibility.isSubmitting,
+});
+
+export default connect(mapStateToProps, mapDispatchToProps)(ExampleForm);
+exports.ExampleFormPure = ExampleForm; // unconnected for enzyme tests
```
Comment (omgaz) [inline src/examples/forms.js]: enzyme isnt relavant just say unit

File: src/helpers/reactSelectAllowCreateOverride.js [hunk 11]
```diff
@@ -0,0 +1,55 @@
+/*eslint-disable */
```
Comment (omgaz) [inline src/helpers/reactSelectAllowCreateOverride.js]: Kill it

**Comment by vinteo** [comment]
LGTM

**Comment by omgaz** [comment]
I'll just double check the versions in the morning and test with @vinteo in the morning. Then we can merge and release for @etaoins.

**Comment by davepoon** [comment]
LGTM :thumbsup:

**Comment by etaoins** [comment]
LGTM :+1:

**Comment by omgaz** [comment]
@vinteo, @davepoon, @etaoins merge?

**Comment by etaoins** [comment]
:rocket:

**Pull request merged**